DuckDuckGo/Onboarding/ContextualOnboarding/ContextualOnboardingStateMachine.swift

```diff
@@ -0,0 +1,260 @@
+            }
+        }
+    }
+
+    var state: ContextualOnboardingState {
+        get {
+            return ContextualOnboardingState(rawValue: stateString) ?? .onboardingCompleted
+        }
+        set {
+            stateString = newValue.rawValue
+        }
+    }
+
+    var lastVisitTab: Tab?
+    var lastVisitSite: URL?
+
+    init(trackerMessageProvider: TrackerMessageProviding = TrackerMessageProvider(),
+         startupPreferences: StartupPreferences = StartupPreferences.shared) {
+        self.trackerMessageProvider = trackerMessageProvider
+        self.startUpPreferences = startupPreferences
+    }
+
+    func dialogTypeForTab(_ tab: Tab, privacyInfo: PrivacyInfo? = nil) -> ContextualDialogType? {
+        let info = privacyInfo ?? tab.privacyInfo
+        guard case .url = tab.content else {
+            return nil
+        }
+        guard let url = tab.url else { return nil }
+
+        if lastVisitTab != nil && tab != lastVisitTab && url == URL.duckDuckGo && state != .showFireButton  {
+            lastVisitTab = tab
+            lastVisitSite = url
+            return nil
+        }
+        lastVisitTab = tab
+        lastVisitSite = url
+        if url.isDuckDuckGoSearch {
+            return dialogPerSearch()
+        } else {
+            return dialogPerSiteVisit(privacyInfo: info)
+        }
+    }
+
+    private func dialogPerSearch() -> ContextualDialogType? {
+        switch state {
+        case .showSearchDone, .fireUsedShowSearchDone:
+            return .searchDone(shouldFollowUp: true)
+        case .showBlockedTrackers, .showMajorOrNoTracker:
+            return .searchDone(shouldFollowUp: false)
+        case .showTryASite:
+            return .tryASite
+        case .showFireButton:
+            return .tryFireButton
+        case .showHighFive:
+            return .highFive
+        default:
+            return nil
+        }
+    }
+
+    private func dialogPerSiteVisit(privacyInfo: PrivacyInfo?) -> ContextualDialogType? {
+        switch state {
+        case .showTryASearch:
+            return .tryASearch
+        case .showBlockedTrackers, .showMajorOrNoTracker, .searchDoneShowBlockedTrackers, .searchDoneShowMajorOrNoTracker, .fireUsedShowSearchDone:
+            guard let privacyInfo else { return nil }
+            guard let message = trackerMessageProvider.trackerMessage(privacyInfo: privacyInfo) else { return nil }
+            return .trackers(message: message, shouldFollowUp: true)
+        case .showFireButton:
+            return .tryFireButton
+        case .showHighFive:
+            return .highFive
+        default:
+            return nil
+        }
+
+    }
+
+    func updateStateFor(tab: Tab) {
+        guard case .url = tab.content else {
+            return
+        }
+        guard let url = tab.url else { return }
+
+        if lastVisitTab != nil && tab != lastVisitTab && url == URL.duckDuckGo && state != .showFireButton  {
```

Comment:
same as this comment https://github.com/duckduckgo/macos-browser/pull/3285/files#r1796776389